### ISSUE TITLE
clatd/Tayga fails to receive/route packets when other firewall rules are present (e.g. UFW, Docker networking)

### ISSUE THREAD

**Issue by jivanpal** [toreanderson/clatd#42]
Not a duplicate of #28.

On a Debian 12 system, when UFW is enabled or Docker is running, Tayga no longer functions properly; all outgoing IPv4 packets get dropped somewhere. ICMP pings using `ping` seemingly succeed, but this is some sort of false positive, as nothing is actually received by the other host (as can be seen by packet sniffing on the wire), and the host running clatd can see problems itself through failed traceroutes and TCP/UDP exchanges.

Debugging logs for clatd and Tayga don't show anything different between the two situations. Tayga does not provide any additional information that might be useful when run manually after manually setting up the clatd interface and routes. Rather, Tayga just listens quietly and doesn't provide any debugging logs.

The iptables tables don't show anything unusual. When UFW is installed and has been enabled at least once, the UFW chains and rules are present. When Docker is running, the DOCKER chain is present with its DNAT rules. UFW and Docker continue working properly.

**Comment by jivanpal** [comment]
Correction: all ICMP traffic seems to be unaffected, but TCP and UDP traffic is dropped.

**Comment by toreanderson** [comment]
Based on your description, it seems highly llikely that UFW and/or Docker is inserting firewall rules or something like that that cause traffic forwarded between `clat-dev` and `plat-dev` to be dropped. 

Could you send me some more debugging output please?

1. Output from starting clatd in debug mode (`clatd -d`)
2. Output from `iptables-save`, `ip6tables-save` and `nft list ruleset` when the problem is in effect
3. Simultaneous packet captures on `clat-dev` and `plat-dev`. For example, `tcpdump -i clat` and `tcpdump -i eth0 net 64:ff9b::/96` - replacing «clat» with your `clat-dev`, «eth0» with your `plat-dev` and «64:ff9b::/96» with your `plat-prefix`.
4. While the packet captures are running, send some test traffic (e.g., `dig +vc @1.1.1.1 . SOA`).

**Comment by jivanpal** [comment]
Disabled UFW, ensured that IPv4/DHCPv4 is disabled in Netplan configuration, disabled clatd Systemd unit, rebooted.

Kernel version (`uname -a`):

```
Linux mail 6.1.0-28-cloud-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.119-1 (2024-11-22) x86_64 GNU/Linux
```

Netplan config (`/etc/netplan/50-cloud-init.yaml`):

```
network:
    ethernets:
        enX0:
            dhcp4: false
    version: 2
```

Output of `ip a` before starting clatd:

```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host noprefixroute 
       valid_lft forever preferred_lft forever
2: enX0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 6e:fd:ca:97:68:2b brd ff:ff:ff:ff:ff:ff
    inet6 fd41:b008:2015:0:6cfd:caff:fe97:682b/64 scope global dynamic mngtmpaddr noprefixroute 
       valid_lft 43185sec preferred_lft 43185sec
    inet6 2a02:6b6f:fc22:4c00:6cfd:caff:fe97:682b/64 scope global dynamic mngtmpaddr noprefixroute 
       valid_lft 43185sec preferred_lft 43185sec
    inet6 fe80::6cfd:caff:fe97:682b/64 scope link 
       valid_lft forever preferred_lft forever
```

<details>
<summary>Output of <code>clatd -d</code>, there is no clatd config file:</summary>

```
Configuration successfully read, dumping it:
  clat-dev=clat
  clat-v4-addr=192.0.0.1
  clat-v6-addr=<undefined>
  cmd-ip=ip
  cmd-ip6tables=ip6tables
  cmd-tayga=tayga
  debug=1
  dns64-servers=<undefined>
  forwarding-enable=1
  ip6tables-enable=<undefined>
  plat-dev=<undefined>
  plat-fallback-prefix=<undefined>
  plat-prefix=<undefined>
  proxynd-enable=1
  quiet=0
  script-down=<undefined>
  script-up=<undefined>
  tayga-conffile=<undefined>
  tayga-v4-addr=192.0.0.2
  v4-conncheck-delay=10
  v4-conncheck-enable=1
  v4-defaultroute-advmss=0
  v4-defaultroute-enable=1
  v4-defaultroute-metric=2048
  v4-defaultroute-mtu=1260
  v4-defaultroute-replace=0
Starting clatd v1.6nat46 by Tore Anderson <tore@fud.no>
Performing DNS64-based PLAT prefix discovery (cf. RFC 7050)
Looking up 'ipv4only.arpa' using system resolver
check_wka(): Testing to see if 64:ff9b:0:0:0:0:c000:ab was DNS64-synthesised
Inferred PLAT prefix 64:ff9b::/96 from AAAA record 64:ff9b:0:0:0:0:c000:ab
check_wka(): Testing to see if 64:ff9b:0:0:0:0:c000:aa was DNS64-synthesised
Inferred PLAT prefix 64:ff9b::/96 from AAAA record 64:ff9b:0:0:0:0:c000:aa
Using PLAT (NAT64) prefix: 64:ff9b::/96
get_plat_dev(): finding which network dev faces the PLAT
get_plat_dev(): Found PLAT-facing device: enX0
Device facing the PLAT: enX0
Attempting to derive a CLAT IPv6 address from an IPv6 address on 'enX0'
Using CLAT IPv4 address: 192.0.0.1
Using CLAT IPv6 address: 2a02:6b6f:fc22:4c00:6cfd:cac1:a797:682b
Checking if this system already has IPv4 connectivity in 10 sec(s)
Reading sysctl /proc/sys/net/ipv6/conf/all/forwarding
/proc/sys/net/ipv6/conf/all/forwarding is set to '0'
Enabling IPv6 forwarding
Reading sysctl /proc/sys/net/ipv6/conf/default/accept_ra
/proc/sys/net/ipv6/conf/default/accept_ra is set to '1'
Changing /proc/sys/net/ipv6/conf/default/accept_ra from 1 to 2 to prevent connectivity loss after enabling IPv6 forwarding
Setting sysctl /proc/sys/net/ipv6/conf/default/accept_ra=2
Reading sysctl /proc/sys/net/ipv6/conf/enX0/accept_ra
/proc/sys/net/ipv6/conf/enX0/accept_ra is set to '0'
Reading sysctl /proc/sys/net/ipv6/conf/lo/accept_ra
/proc/sys/net/ipv6/conf/lo/accept_ra is set to '1'
Changing /proc/sys/net/ipv6/conf/lo/accept_ra from 1 to 2 to prevent connectivity loss after enabling IPv6 forwarding
Setting sysctl /proc/sys/net/ipv6/conf/lo/accept_ra=2
Setting sysctl /proc/sys/net/ipv6/conf/all/forwarding=1
Enabling Proxy-ND for 2a02:6b6f:fc22:4c00:6cfd:cac1:a797:682b on enX0
Reading sysctl /proc/sys/net/ipv6/conf/enX0/proxy_ndp
/proc/sys/net/ipv6/conf/enX0/proxy_ndp is set to '0'
Setting sysctl /proc/sys/net/ipv6/conf/enX0/proxy_ndp=1
Enabled Proxy-ND sysctl for enX0
cmd(ip -6 neighbour add proxy 2a02:6b6f:fc22:4c00:6cfd:cac1:a797:682b dev enX0)
Creating and configuring up CLAT device 'clat'
cmd(tayga --config /tmp/0KsrQehfdY --mktun -d)
Created persistent tun device clat
cmd(ip link set up dev clat)
cmd(ip -4 address add 192.0.0.1 dev clat)
cmd(ip -6 route add 2a02:6b6f:fc22:4c00:6cfd:cac1:a797:682b dev clat)
Adding IPv4 default route via the CLAT
cmd(ip -4 route add default dev clat metric 2048 mtu 1260 advmss 1220)
Starting up TAYGA, using config file '/tmp/0KsrQehfdY'
cmd(tayga --config /tmp/0KsrQehfdY --nodetach -d)
starting TAYGA 0.9.2
Using tun device clat with MTU 1500
TAYGA's IPv4 address: 192.0.0.2
TAYGA's IPv6 address: 64:ff9b::c000:2
NAT64 prefix: 64:ff9b::/96
Note: traffic between IPv6 hosts and private IPv4 addresses (i.e. to/from 64:ff9b::10.0.0.0/104, 64:ff9b::192.168.0.0/112, etc) will be dropped.  Use a translation prefix within your organization's IPv6 address space instead of 64:ff9b::/96 if you need your IPv6 hosts to communicate with private IPv4 addresses.
```

</details>

clatd continues to run in the background in another shell.

Output of `ip a` after clatd has successfully started:

```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host noprefixroute 
       valid_lft forever preferred_lft forever
2: enX0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether 6e:fd:ca:97:68:2b brd ff:ff:ff:ff:ff:ff
    inet6 fd41:b008:2015:0:6cfd:caff:fe97:682b/64 scope global dynamic mngtmpaddr noprefixroute 
       valid_lft 43102sec preferred_lft 43102sec
    inet6 2a02:6b6f:fc22:4c00:6cfd:caff:fe97:682b/64 scope global dynamic mngtmpaddr noprefixroute 
       valid_lft 43102sec preferred_lft 43102sec
    inet6 fe80::6cfd:caff:fe97:682b/64 scope link 
       valid_lft forever preferred_lft forever
3: clat: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 500
    link/none 
    inet 192.0.0.1/32 scope global clat
       valid_lft forever preferred_lft forever
    inet6 fe80::d5a:6430:b63d:bcf5/64 scope link stable-privacy 
       valid_lft forever preferred_lft forever
```

IPv4 connectivity works fine at this point, e.g. I can access <http://8.8.8.8/> using Curl.

***

UFW is now enabled with one firewall rule. Output of `ufw status verbose` after enabling:

```
Status: active
Logging: on (low)
Default: deny (incoming), allow (outgoing), deny (routed)
New profiles: skip

To                         Action      From
--                         ------      ----
22/tcp (OpenSSH (v6))      ALLOW IN    Anywhere (v6)             
```

<details>
<summary>Output of <code>iptables-save</code>:</summary>

```
# Generated by iptables-save v1.8.9 (nf_tables) on Sat Dec 21 22:55:00 2024
*filter
:INPUT DROP [0:0]
:FORWARD DROP [0:0]
:OUTPUT ACCEPT [0:0]
:ufw-after-forward - [0:0]
:ufw-after-input - [0:0]
:ufw-after-logging-forward - [0:0]
:ufw-after-logging-input - [0:0]
:ufw-after-logging-output - [0:0]
:ufw-after-output - [0:0]
:ufw-before-forward - [0:0]
:ufw-before-input - [0:0]
:ufw-before-logging-forward - [0:0]
:ufw-before-logging-input - [0:0]
:ufw-before-logging-output - [0:0]
:ufw-before-output - [0:0]
:ufw-logging-allow - [0:0]
:ufw-logging-deny - [0:0]
:ufw-not-local - [0:0]
:ufw-reject-forward - [0:0]
:ufw-reject-input - [0:0]
:ufw-reject-output - [0:0]
:ufw-skip-to-policy-forward - [0:0]
:ufw-skip-to-policy-input - [0:0]
:ufw-skip-to-policy-output - [0:0]
:ufw-track-forward - [0:0]
:ufw-track-input - [0:0]
:ufw-track-output - [0:0]
:ufw-user-forward - [0:0]
:ufw-user-input - [0:0]
:ufw-user-limit - [0:0]
:ufw-user-limit-accept - [0:0]
:ufw-user-logging-forward - [0:0]
:ufw-user-logging-input - [0:0]
:ufw-user-logging-output - [0:0]
:ufw-user-output - [0:0]
-A INPUT -j ufw-before-logging-input
-A INPUT -j ufw-before-input
-A INPUT -j ufw-after-input
-A INPUT -j ufw-after-logging-input
-A INPUT -j ufw-reject-input
-A INPUT -j ufw-track-input
-A FORWARD -j ufw-before-logging-forward
-A FORWARD -j ufw-before-forward
-A FORWARD -j ufw-after-forward
-A FORWARD -j ufw-after-logging-forward
-A FORWARD -j ufw-reject-forward
-A FORWARD -j ufw-track-forward
-A OUTPUT -j ufw-before-logging-output
-A OUTPUT -j ufw-before-output
-A OUTPUT -j ufw-after-output
-A OUTPUT -j ufw-after-logging-output
-A OUTPUT -j ufw-reject-output
-A OUTPUT -j ufw-track-output
-A ufw-after-input -p udp -m udp --dport 137 -j ufw-skip-to-policy-input
-A ufw-after-input -p udp -m udp --dport 138 -j ufw-skip-to-policy-input
-A ufw-after-input -p tcp -m tcp --dport 139 -j ufw-skip-to-policy-input
-A ufw-after-input -p tcp -m tcp --dport 445 -j ufw-skip-to-policy-input
-A ufw-after-input -p udp -m udp --dport 67 -j ufw-skip-to-policy-input
-A ufw-after-input -p udp -m udp --dport 68 -j ufw-skip-to-policy-input
-A ufw-after-input -m addrtype --dst-type BROADCAST -j ufw-skip-to-policy-input
-A ufw-after-logging-forward -m limit --limit 3/min --limit-burst 10 -j LOG --log-prefix "[UFW BLOCK] "
-A ufw-after-logging-input -m limit --limit 3/min --limit-burst 10 -j LOG --log-prefix "[UFW BLOCK] "
-A ufw-before-forward -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A ufw-before-forward -p icmp -m icmp --icmp-type 3 -j ACCEPT
-A ufw-before-forward -p icmp -m icmp --icmp-type 11 -j ACCEPT
-A ufw-before-forward -p icmp -m icmp --icmp-type 12 -j ACCEPT
-A ufw-before-forward -p icmp -m icmp --icmp-type 8 -j ACCEPT
-A ufw-before-forward -j ufw-user-forward
-A ufw-before-input -i lo -j ACCEPT
-A ufw-before-input -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A ufw-before-input -m conntrack --ctstate INVALID -j ufw-logging-deny
-A ufw-before-input -m conntrack --ctstate INVALID -j DROP
-A ufw-before-input -p icmp -m icmp --icmp-type 3 -j ACCEPT
-A ufw-before-input -p icmp -m icmp --icmp-type 11 -j ACCEPT
-A ufw-before-input -p icmp -m icmp --icmp-type 12 -j ACCEPT
-A ufw-before-input -p icmp -m icmp --icmp-type 8 -j ACCEPT
-A ufw-before-input -p udp -m udp --sport 67 --dport 68 -j ACCEPT
-A ufw-before-input -j ufw-not-local
-A ufw-before-input -d 224.0.0.251/32 -p udp -m udp --dport 5353 -j ACCEPT
-A ufw-before-input -d 239.255.255.250/32 -p udp -m udp --dport 1900 -j ACCEPT
-A ufw-before-input -j ufw-user-input
-A ufw-before-output -o lo -j ACCEPT
-A ufw-before-output -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A ufw-before-output -j ufw-user-output
-A ufw-logging-allow -m limit --limit 3/min --limit-burst 10 -j LOG --log-prefix "[UFW ALLOW] "
-A ufw-logging-deny -m conntrack --ctstate INVALID -m limit --limit 3/min --limit-burst 10 -j RETURN
-A ufw-logging-deny -m limit --limit 3/min --limit-burst 10 -j LOG --log-prefix "[UFW BLOCK] "
-A ufw-not-local -m addrtype --dst-type LOCAL -j RETURN
-A ufw-not-local -m addrtype --dst-type MULTICAST -j RETURN
-A ufw-not-local -m addrtype --dst-type BROADCAST -j RETURN
-A ufw-not-local -m limit --limit 3/min --limit-burst 10 -j ufw-logging-deny
-A ufw-not-local -j DROP
-A ufw-skip-to-policy-forward -j DROP
-A ufw-skip-to-policy-input -j DROP
-A ufw-skip-to-policy-output -j ACCEPT
-A ufw-track-output -p tcp -m conntrack --ctstate NEW -j ACCEPT
-A ufw-track-output -p udp -m conntrack --ctstate NEW -j ACCEPT
-A ufw-user-limit -m limit --limit 3/min -j LOG --log-prefix "[UFW LIMIT BLOCK] "
-A ufw-user-limit -j REJECT --reject-with icmp-port-unreachable
-A ufw-user-limit-accept -j ACCEPT
COMMIT
# Completed on Sat Dec 21 22:55:00 2024
```

</details>

<details>
<summary>Output of <code>ip6tables-save</code>:</summary>

```
# Generated by ip6tables-save v1.8.9 (nf_tables) on Sat Dec 21 22:55:11 2024
*filter
:INPUT DROP [29:4991]
:FORWARD DROP [2:160]
:OUTPUT ACCEPT [0:0]
:ufw6-after-forward - [0:0]
:ufw6-after-input - [0:0]
:ufw6-after-logging-forward - [0:0]
:ufw6-after-logging-input - [0:0]
:ufw6-after-logging-output - [0:0]
:ufw6-after-output - [0:0]
:ufw6-before-forward - [0:0]
:ufw6-before-input - [0:0]
:ufw6-before-logging-forward - [0:0]
:ufw6-before-logging-input - [0:0]
:ufw6-before-logging-output - [0:0]
:ufw6-before-output - [0:0]
:ufw6-logging-allow - [0:0]
:ufw6-logging-deny - [0:0]
:ufw6-reject-forward - [0:0]
:ufw6-reject-input - [0:0]
:ufw6-reject-output - [0:0]
:ufw6-skip-to-policy-forward - [0:0]
:ufw6-skip-to-policy-input - [0:0]
:ufw6-skip-to-policy-output - [0:0]
:ufw6-track-forward - [0:0]
:ufw6-track-input - [0:0]
:ufw6-track-output - [0:0]
:ufw6-user-forward - [0:0]
:ufw6-user-input - [0:0]
:ufw6-user-limit - [0:0]
:ufw6-user-limit-accept - [0:0]
:ufw6-user-logging-forward - [0:0]
:ufw6-user-logging-input - [0:0]
:ufw6-user-logging-output - [0:0]
:ufw6-user-output - [0:0]
-A INPUT -j ufw6-before-logging-input
-A INPUT -j ufw6-before-input
-A INPUT -j ufw6-after-input
-A INPUT -j ufw6-after-logging-input
-A INPUT -j ufw6-reject-input
-A INPUT -j ufw6-track-input
-A FORWARD -j ufw6-before-logging-forward
-A FORWARD -j ufw6-before-forward
-A FORWARD -j ufw6-after-forward
-A FORWARD -j ufw6-after-logging-forward
-A FORWARD -j ufw6-reject-forward
-A FORWARD -j ufw6-track-forward
-A OUTPUT -j ufw6-before-logging-output
-A OUTPUT -j ufw6-before-output
-A OUTPUT -j ufw6-after-output
-A OUTPUT -j ufw6-after-logging-output
-A OUTPUT -j ufw6-reject-output
-A OUTPUT -j ufw6-track-output
-A ufw6-after-input -p udp -m udp --dport 137 -j ufw6-skip-to-policy-input
-A ufw6-after-input -p udp -m udp --dport 138 -j ufw6-skip-to-policy-input
-A ufw6-after-input -p tcp -m tcp --dport 139 -j ufw6-skip-to-policy-input
-A ufw6-after-input -p tcp -m tcp --dport 445 -j ufw6-skip-to-policy-input
-A ufw6-after-input -p udp -m udp --dport 546 -j ufw6-skip-to-policy-input
-A ufw6-after-input -p udp -m udp --dport 547 -j ufw6-skip-to-policy-input
-A ufw6-after-logging-forward -m limit --limit 3/min --limit-burst 10 -j LOG --log-prefix "[UFW BLOCK] "
-A ufw6-after-logging-input -m limit --limit 3/min --limit-burst 10 -j LOG --log-prefix "[UFW BLOCK] "
-A ufw6-before-forward -m rt --rt-type 0 -j DROP
-A ufw6-before-forward -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A ufw6-before-forward -p ipv6-icmp -m icmp6 --icmpv6-type 1 -j ACCEPT
-A ufw6-before-forward -p ipv6-icmp -m icmp6 --icmpv6-type 2 -j ACCEPT
-A ufw6-before-forward -p ipv6-icmp -m icmp6 --icmpv6-type 3 -j ACCEPT
-A ufw6-before-forward -p ipv6-icmp -m icmp6 --icmpv6-type 4 -j ACCEPT
-A ufw6-before-forward -p ipv6-icmp -m icmp6 --icmpv6-type 128 -j ACCEPT
-A ufw6-before-forward -p ipv6-icmp -m icmp6 --icmpv6-type 129 -j ACCEPT
-A ufw6-before-forward -j ufw6-user-forward
-A ufw6-before-input -i lo -j ACCEPT
-A ufw6-before-input -m rt --rt-type 0 -j DROP
-A ufw6-before-input -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 129 -j ACCEPT
-A ufw6-before-input -m conntrack --ctstate INVALID -j ufw6-logging-deny
-A ufw6-before-input -m conntrack --ctstate INVALID -j DROP
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 1 -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 2 -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 3 -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 4 -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 128 -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 133 -m hl --hl-eq 255 -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 134 -m hl --hl-eq 255 -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 135 -m hl --hl-eq 255 -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 136 -m hl --hl-eq 255 -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 141 -m hl --hl-eq 255 -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 142 -m hl --hl-eq 255 -j ACCEPT
-A ufw6-before-input -s fe80::/10 -p ipv6-icmp -m icmp6 --icmpv6-type 130 -j ACCEPT
-A ufw6-before-input -s fe80::/10 -p ipv6-icmp -m icmp6 --icmpv6-type 131 -j ACCEPT
-A ufw6-before-input -s fe80::/10 -p ipv6-icmp -m icmp6 --icmpv6-type 132 -j ACCEPT
-A ufw6-before-input -s fe80::/10 -p ipv6-icmp -m icmp6 --icmpv6-type 143 -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 148 -m hl --hl-eq 255 -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 149 -m hl --hl-eq 255 -j ACCEPT
-A ufw6-before-input -s fe80::/10 -p ipv6-icmp -m icmp6 --icmpv6-type 151 -m hl --hl-eq 1 -j ACCEPT
-A ufw6-before-input -s fe80::/10 -p ipv6-icmp -m icmp6 --icmpv6-type 152 -m hl --hl-eq 1 -j ACCEPT
-A ufw6-before-input -s fe80::/10 -p ipv6-icmp -m icmp6 --icmpv6-type 153 -m hl --hl-eq 1 -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 144 -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 145 -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 146 -j ACCEPT
-A ufw6-before-input -p ipv6-icmp -m icmp6 --icmpv6-type 147 -j ACCEPT
-A ufw6-before-input -s fe80::/10 -d fe80::/10 -p udp -m udp --sport 547 --dport 546 -j ACCEPT
-A ufw6-before-input -d ff02::fb/128 -p udp -m udp --dport 5353 -j ACCEPT
-A ufw6-before-input -d ff02::f/128 -p udp -m udp --dport 1900 -j ACCEPT
-A ufw6-before-input -j ufw6-user-input
-A ufw6-before-output -o lo -j ACCEPT
-A ufw6-before-output -m rt --rt-type 0 -j DROP
-A ufw6-before-output -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A ufw6-before-output -p ipv6-icmp -m icmp6 --icmpv6-type 1 -j ACCEPT
-A ufw6-before-output -p ipv6-icmp -m icmp6 --icmpv6-type 2 -j ACCEPT
-A ufw6-before-output -p ipv6-icmp -m icmp6 --icmpv6-type 3 -j ACCEPT
-A ufw6-before-output -p ipv6-icmp -m icmp6 --icmpv6-type 4 -j ACCEPT
-A ufw6-before-output -p ipv6-icmp -m icmp6 --icmpv6-type 128 -j ACCEPT
-A ufw6-before-output -p ipv6-icmp -m icmp6 --icmpv6-type 129 -j ACCEPT
-A ufw6-before-output -p ipv6-icmp -m icmp6 --icmpv6-type 133 -m hl --hl-eq 255 -j ACCEPT
-A ufw6-before-output -p ipv6-icmp -m icmp6 --icmpv6-type 136 -m hl --hl-eq 255 -j ACCEPT
-A ufw6-before-output -p ipv6-icmp -m icmp6 --icmpv6-type 135 -m hl --hl-eq 255 -j ACCEPT
-A ufw6-before-output -p ipv6-icmp -m icmp6 --icmpv6-type 134 -m hl --hl-eq 255 -j ACCEPT
-A ufw6-before-output -p ipv6-icmp -m icmp6 --icmpv6-type 141 -m hl --hl-eq 255 -j ACCEPT
-A ufw6-before-output -p ipv6-icmp -m icmp6 --icmpv6-type 142 -m hl --hl-eq 255 -j ACCEPT
-A ufw6-before-output -s fe80::/10 -p ipv6-icmp -m icmp6 --icmpv6-type 130 -j ACCEPT
-A ufw6-before-output -s fe80::/10 -p ipv6-icmp -m icmp6 --icmpv6-type 131 -j ACCEPT
-A ufw6-before-output -s fe80::/10 -p ipv6-icmp -m icmp6 --icmpv6-type 132 -j ACCEPT
-A ufw6-before-output -s fe80::/10 -p ipv6-icmp -m icmp6 --icmpv6-type 143 -j ACCEPT
-A ufw6-before-output -p ipv6-icmp -m icmp6 --icmpv6-type 148 -m hl --hl-eq 255 -j ACCEPT
-A ufw6-before-output -p ipv6-icmp -m icmp6 --icmpv6-type 149 -m hl --hl-eq 255 -j ACCEPT
-A ufw6-before-output -s fe80::/10 -p ipv6-icmp -m icmp6 --icmpv6-type 151 -m hl --hl-eq 1 -j ACCEPT
-A ufw6-before-output -s fe80::/10 -p ipv6-icmp -m icmp6 --icmpv6-type 152 -m hl --hl-eq 1 -j ACCEPT
-A ufw6-before-output -s fe80::/10 -p ipv6-icmp -m icmp6 --icmpv6-type 153 -m hl --hl-eq 1 -j ACCEPT
-A ufw6-before-output -j ufw6-user-output
-A ufw6-logging-allow -m limit --limit 3/min --limit-burst 10 -j LOG --log-prefix "[UFW ALLOW] "
-A ufw6-logging-deny -m conntrack --ctstate INVALID -m limit --limit 3/min --limit-burst 10 -j RETURN
-A ufw6-logging-deny -m limit --limit 3/min --limit-burst 10 -j LOG --log-prefix "[UFW BLOCK] "
-A ufw6-skip-to-policy-forward -j DROP
-A ufw6-skip-to-policy-input -j DROP
-A ufw6-skip-to-policy-output -j ACCEPT
-A ufw6-track-output -p tcp -m conntrack --ctstate NEW -j ACCEPT
-A ufw6-track-output -p udp -m conntrack --ctstate NEW -j ACCEPT
-A ufw6-user-input -p tcp -m tcp --dport 22 -m comment --comment "\'dapp_OpenSSH\'" -j ACCEPT
-A ufw6-user-limit -m limit --limit 3/min -j LOG --log-prefix "[UFW LIMIT BLOCK] "
-A ufw6-user-limit -j REJECT --reject-with icmp6-port-unreachable
-A ufw6-user-limit-accept -j ACCEPT
COMMIT
# Completed on Sat Dec 21 22:55:11 2024
```

</details>

`nft` is not installed on the system.

***

Performed packet trace whilst running `host -t SOA . 1.1.1.1`.

Output of `host` command:

```
; communications error to 1.1.1.1#53: timed out
;; communications error to 1.1.1.1#53: timed out
;; no servers could be reached
```

Output of `tcpdump -i clat`:

```
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on clat, link-type RAW (Raw IP), snapshot length 262144 bytes
22:57:10.952040 IP mail.51891 > one.one.one.one.domain: 50724+ SOA? . (17)
22:57:10.952102 IP6 2a02:6b6f:fc22:4c00:6cfd:cac1:a797:682b.51891 > 64:ff9b::101:101.domain: 50724+ SOA? . (17)
22:57:15.956606 IP mail.48821 > one.one.one.one.domain: 50724+ SOA? . (17)
22:57:15.956743 IP6 2a02:6b6f:fc22:4c00:6cfd:cac1:a797:682b.48821 > 64:ff9b::101:101.domain: 50724+ SOA? . (17)
```

Output of `tcpdump -i enX0 net 64:ff9b::/96`:

```
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on enX0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
```

That is, no such packets sent/received on `enX0`.

**Comment by toreanderson** [comment]
So what I can tell from this is that the firewall ruleset loaded by UFW seems to block the forwarding of IPv6 packets between the `clat` and `enX0` devices. That will certainly break things.

I see that `ufw status verbose` says `Default: (…) deny (routed)` which probably means that UFW is set up to block all forwarding by default. This seems to match the actual rules in your ruleset:

```
:FORWARD DROP [2:160]
```

I think that you might see in your kernel messages (check `dmesg` or `journalctl -k`) `[UFW BLOCK]` messages whenever you try to send traffic through the CLAT. These would be generated by the following ip6tables rule:

```
-A ufw6-after-logging-forward -m limit --limit 3/min --limit-burst 10 -j LOG --log-prefix "[UFW BLOCK] "
```

I am a little bit surprised that `clatd` didn't try to inject ip6tables rules on its own. It is supposed to do so if it sees that ip6tables is enabled on the system (by checking whether or not `/sys/module/ip6table_filter` exists). I suppose it is possible that the kernel module has been renamed. Could you try starting `clatd` with the `ip6tables-enable=1` command line option (or add it to clatd.conf), and see if that works better?

Note that there is no guarantee that UFW won't overwrite any rules added by `clatd`. If it does so, you might want to figure out how to make UFW allow routed IPv6 traffic between the `clat` and `enX0` interfaces. You could then add those UFW commands to `post-up` and `post-down` so that they are added/removed by clatd on demand.

Do let me know if you could figure that out, as I can imagine others might run across the same problem.

**Comment by jivanpal** [comment]
UFW is/was indeed logging the following (the change in CLAT IPv6 address compared to the original post is due to a change in MAC address, as I'm now testing on a different VM):

```
[UFW BLOCK] IN=clat OUT=enX0 MAC= SRC=2a02:6b6f:fc22:4c00:d438:0fc1:a7d7:e269 DST=0064:ff9b:0000:0000:0000:0000:0808:0808 LEN=82 TC=0 HOPLIMIT=62 FLOWLBL=0 PROTO=UDP SPT=49828 DPT=53 LEN=42
```

However, this confuses me: why are such packets (IPv6 packets with source address being that of the CLAT (`*c1:a7*`) and destination address being the IPv4 destination's IPv4-mapped IPv6 address (`64:ff9b:*`)) being sent out of enX0 towards the clat interface, rather than the interfaces or addresses being vice-versa? In any case, adding the following rules to UFW causes no change in behaviour, which I also find strange:

```
$ sudo ufw allow in on clat
$ sudo ufw allow in on enX0

$ sudo ufw status verbose
Status: active
Logging: on (low)
Default: deny (incoming), allow (outgoing), deny (routed)
New profiles: skip

To                         Action      From
--                         ------      ----
Anywhere on clat           ALLOW IN    Anywhere                  
Anywhere on enX0           ALLOW IN    Anywhere                  
22/tcp (OpenSSH (v6))      ALLOW IN    Anywhere (v6)             
Anywhere (v6) on clat      ALLOW IN    Anywhere (v6)             
Anywhere (v6) on enX0      ALLOW IN    Anywhere (v6)             
```

***

Reverting back to the original UFW rules, adding `ip6tables-enable=1` to`/etc/clatd.conf` did in fact fix the issue, with the addition of one other UFW rule:

```
$ sudo ufw allow in on clat from 192.0.0.1

$ sudo ufw status verbose
Status: active
Logging: on (low)
Default: deny (incoming), allow (outgoing), deny (routed)
New profiles: skip

To                         Action      From
--                         ------      ----
Anywhere on clat           ALLOW IN    192.0.0.1                 
22/tcp (OpenSSH (v6))      ALLOW IN    Anywhere (v6)
```

Without that additional rule, UFW logs the following:

```
[UFW BLOCK] IN=clat OUT= MAC= SRC=192.0.0.1 DST=224.0.0.252 LEN=56 TOS=0x00 PREC=0x00 TTL=255 ID=5455 PROTO=UDP SPT=5355 DPT=5355 LEN=36
```

I'll report back here if anything stops working in the future, but so far, it seems to continue working even after other UFW rules are added or removed, or UFW is disabled and re-enabled. I'm happy for you to close this issue if you think that's appropriate. I still haven't tested this with Docker (whether with or without UFW), but if I have problems with it once I do test it, I'd be happy to open a separate issue for that.

***

As for why `ip6tables-enable=1` wasn't set automatically...

The following packages are installed:

```
$ dpkg -l | grep -ie netfilter -e table
ii  iptables                         1.8.9-2                        amd64        administration tools for packet filtering and NAT
ii  kbd                              2.5.1-1+b1                     amd64        Linux console font and keytable utilities
ii  libevent-core-2.1-7:amd64        2.1.12-stable-8                amd64        Asynchronous event notification library (core)
ii  libip4tc2:amd64                  1.8.9-2                        amd64        netfilter libip4tc library
ii  libip6tc2:amd64                  1.8.9-2                        amd64        netfilter libip6tc library
ii  libnetfilter-conntrack3:amd64    1.0.9-3                        amd64        Netfilter netlink-conntrack library
ii  libnfnetlink0:amd64              1.0.2-2                        amd64        Netfilter netlink library
ii  libnftnl11:amd64                 1.2.4-2                        amd64        Netfilter nftables userspace API library
ii  libxtables12:amd64               1.8.9-2                        amd64        netfilter xtables library
ii  python3-netifaces:amd64          0.11.0-2+b1                    amd64        portable network interface information - Python 3.x
ii  python3-pyrsistent:amd64         0.18.1-1+b3                    amd64        persistent/functional/immutable data structures for Python
ii  python3-rich                     13.3.1-1                       all          render rich text, tables, progress bars, syntax highlighting, markdown and more
ii  ufw                              0.36.2-1                       all          program for managing a Netfilter firewall
```

The path `/sys/module/ip6table_filter` does not exist. Here is what SysFS looks like on my system, which is generated from the Debian 12 cloud "genericcloud" images (see <https://cloud.debian.org/>):

```
$ sudo find /sys/ -iname '*ip6*'
/sys/kernel/btf/ip6t_rt
/sys/kernel/btf/ip6_tables
/sys/kernel/btf/ip6t_REJECT
/sys/kernel/slab/ip6_dst_cache
/sys/kernel/slab/ip6_mrt_cache
/sys/kernel/slab/ip6-frags
/sys/module/nf_reject_ipv6/holders/ip6t_REJECT
/sys/module/ip6t_rt
/sys/module/ip6_tables
/sys/module/x_tables/holders/ip6t_rt
/sys/module/x_tables/holders/ip6_tables
/sys/module/x_tables/holders/ip6t_REJECT
/sys/module/ip6t_REJECT

$ sudo find /sys/ -iname '*ip6*' -type f
/sys/kernel/btf/ip6t_rt
/sys/kernel/btf/ip6_tables
/sys/kernel/btf/ip6t_REJECT

$ sudo find /sys/ -iname '*ip6*' -type l -exec ls -al {} \;
lrwxrwxrwx 1 root root 0 Jan  7 20:20 /sys/kernel/slab/ip6_dst_cache -> :A-0000256
lrwxrwxrwx 1 root root 0 Jan  7 20:20 /sys/kernel/slab/ip6_mrt_cache -> :0000192
lrwxrwxrwx 1 root root 0 Jan  7 20:20 /sys/kernel/slab/ip6-frags -> :0000184
lrwxrwxrwx 1 root root 0 Jan  7 20:20 /sys/module/nf_reject_ipv6/holders/ip6t_REJECT -> ../../ip6t_REJECT
lrwxrwxrwx 1 root root 0 Jan  7 20:20 /sys/module/x_tables/holders/ip6t_rt -> ../../ip6t_rt
lrwxrwxrwx 1 root root 0 Jan  7 20:20 /sys/module/x_tables/holders/ip6_tables -> ../../ip6_tables
lrwxrwxrwx 1 root root 0 Jan  7 20:20 /sys/module/x_tables/holders/ip6t_REJECT -> ../../ip6t_REJECT

$ sudo find /sys/ -iname '*ip6*' -type d -exec tree -a {} \;
/sys/module/ip6t_rt
├── coresize
├── holders
├── initsize
├── initstate
├── notes
│   ├── .note.gnu.build-id
│   └── .note.Linux
├── refcnt
├── sections
│   ├── .data..read_mostly
│   ├── __dyndbg
│   ├── .exit.data
│   ├── .exit.text
│   ├── .gnu.linkonce.this_module
│   ├── .init.data
│   ├── .init.text
│   ├── __jump_table
│   ├── __mcount_loc
│   ├── .note.gnu.build-id
│   ├── .note.Linux
│   ├── .orc_unwind
│   ├── .orc_unwind_ip
│   ├── .return_sites
│   ├── .rodata
│   ├── .rodata.str1.1
│   ├── .rodata.str1.8
│   ├── .strtab
│   ├── .symtab
│   └── .text
├── taint
└── uevent

4 directories, 28 files
/sys/module/ip6_tables
├── coresize
├── holders
├── initsize
├── initstate
├── notes
│   ├── .note.gnu.build-id
│   └── .note.Linux
├── refcnt
├── sections
│   ├── .altinstr_replacement
│   ├── .altinstructions
│   ├── __bug_table
│   ├── .data
│   ├── .data..read_mostly
│   ├── .exit.data
│   ├── .exit.text
│   ├── .gnu.linkonce.this_module
│   ├── .init.data
│   ├── .init.text
│   ├── __jump_table
│   ├── __kcrctab
│   ├── __kcrctab_gpl
│   ├── __ksymtab
│   ├── __ksymtab_gpl
│   ├── __ksymtab_strings
│   ├── __mcount_loc
│   ├── .note.gnu.build-id
│   ├── .note.Linux
│   ├── .orc_unwind
│   ├── .orc_unwind_ip
│   ├── .retpoline_sites
│   ├── .return_sites
│   ├── .rodata
│   ├── .rodata.str1.1
│   ├── .rodata.str1.8
│   ├── .static_call_sites
│   ├── .strtab
│   ├── .symtab
│   ├── .text
│   └── .text.unlikely
├── taint
└── uevent

4 directories, 39 files
/sys/module/ip6t_REJECT
├── coresize
├── holders
├── initsize
├── initstate
├── notes
│   ├── .note.gnu.build-id
│   └── .note.Linux
├── refcnt
├── sections
│   ├── .data
│   ├── .data..read_mostly
│   ├── .exit.data
│   ├── .exit.text
│   ├── .gnu.linkonce.this_module
│   ├── .init.data
│   ├── .init.text
│   ├── __mcount_loc
│   ├── .note.gnu.build-id
│   ├── .note.Linux
│   ├── .orc_unwind
│   ├── .orc_unwind_ip
│   ├── .return_sites
│   ├── .rodata
│   ├── .rodata.str1.1
│   ├── .rodata.str1.8
│   ├── .strtab
│   ├── .symtab
│   ├── .text
│   └── .text.unlikely
├── taint
└── uevent

4 directories, 28 files
```

***

One other note — You may like to add the following `Before` and `Wants` lines to the default SystemD unit file:

```
[Unit]
Description=464XLAT CLAT daemon
Documentation=man:clatd(8)
After=network-online.target
Before=nss-lookup.target
Wants=network-online.target nss-lookup.target
StartLimitIntervalSec=0
```

This should ensure that clatd only starts if a DNS server is reachable; it has been working reliably for me. Otherwise, if clatd is enabled to start on boot, then if/when it tries to determine the PLAT prefix by querying ipv4only.arpa, the following is likely to happen:

```
20:08:57 clatd-test clatd[485]: Performing DNS64-based PLAT prefix discovery (cf. RFC 7050)
20:10:12 clatd-test clatd[485]: No PLAT prefix could be discovered. Your ISP probably doesn't provide NAT64/DNS64 PLAT service. Exiting.
20:10:12 clatd-test systemd[1]: clatd.service: Deactivated successfully.
```

Manually restarting the service after this timeout gets things working again, but we'd rather start the service at the right time. Alternatively/additionally, we might consider having an extra configuration option to treat lack of discovery of a PLAT prefix not as "NAT64 is not available", but as an error, and then the SystemD unit could use `Restart=on-failure`. This would be useful for servers and other fixed-location machines that are known to be in IPv6-only networks where NAT64 should always be available.

**Comment by toreanderson** [comment]
> However, this confuses me: why are such packets (IPv6 packets with source address being that of the CLAT (*c1:a7*) and destination address being the IPv4 destination's IPv4-mapped IPv6 address (64:ff9b:*)) being sent out of enX0 towards the clat interface, rather than the interfaces or addresses being vice-versa?

Well, it is as expected. TAYGA rewrites your outgoing IPv4 packet so that the source address is its own (`*c1:a7*`) and the destination is IPv4-mapped (`64:ff9b:`). After that is done, it returns the packet to the kernel for routing.

A quick search in the UFW manual tells me your UFW rules are most likely wrong, as they deal with INPUT traffic instead of FORWARD. The right ones might be something like this, could you try them?

```
ufw route allow in on clat out on enX0
ufw route allow in on enX0 out on clat
```

**Comment by jivanpal** [comment]
> Well, it is as expected. TAYGA rewrites your outgoing IPv4 packet so that the source address is its own (*c1:a7*) and the destination is IPv4-mapped (64:ff9b:). After that is done, it returns the packet to the kernel for routing.

Interesting, I obviously don't understand Tayga's intended packet flows, then, because if such packets are really destined for the IPv4 address we're trying to connect to, I would expect those packets to come out of the enX0 interface, but then be routed to the default gateway rather than going into the clat interface. That is, I don't understand why that message says `IN=clat OUT=enX0` rather than just `IN= OUT=enX0`).

> A quick search in the UFW manual tells me your UFW rules are most likely wrong, as they deal with INPUT traffic instead of FORWARD. 

Ah, my mistake! Just the following appears to work, even without `ip6tables-enabled=1` set:

```
ufw route allow in on clat out on enX0
```

I suspect that the other rule is not needed because the IPv4 connections are always initiated by us, so UFW/IPTables permits them as ESTABLISHED. If any application on our host binds to an IPv4 socket/address, then I expect the other rule will also be necessary.

**Comment by toreanderson** [comment]
> Interesting, I obviously don't understand Tayga's intended packet flows, then, because if such packets are really destined for the IPv4 address we're trying to connect to, I would expect those packets to come out of the enX0 interface, but then be routed to another host such as the default gateway rather than going into the clat interface. That is, I don't understand why that message says `IN=clat OUT=enX0` rather than just `IN= OUT=enX0`.

Since TAYGA is a user space process, it cannot route packets directly to `enX0`. So what it does is that it delivers the translated IPv6 packet (i.e., with source `$clat_v6_addr` and destination `$plat-prefix with original IPv4 dst embedded`) back to the Linux kernel via the TUN interface (`clat`).

The kernel will then receive this packet on the `clat` interface, but realise that since the destination IP (`$plat-prefix with original IPv4 dst embedded`) is not present on any local interface, it will process it as a routed packet and perform a routing table lookup for that destination IP. The output interface it finds, is in your case `enX0`. Therefore, the kernel sees this as a packet that came *IN* from the `clat` interface and that is to be routed onwards *OUT* the `enX0` interface. Such packets traverse the `FORWARD` chain of IPTables, ignoring `INPUT`.

> I suspect that the other rule is not needed because the IPv4 connections are always initiated by us, so UFW/IPTables permits them as ESTABLISHED. If any application on our host binds to an IPv4 socket/address, then I expect the other rule will also be necessary.

Agreed.

Hope that settles this issue particular issue, then?

Feel free to open a separate issue for the `ip6tables-enable=1` not working as expected (I bet it is due to the kernel upgrading to NFTables with IPTables compatibility, and that probably has a different module name), and also a MR for the systemd unit dependencies you suggested.

**Comment by jivanpal** [comment]
Thanks for the explanation, this makes sense to me now.

> Feel free to open a separate issue for the `ip6tables-enable=1` not working as expected (I bet it is due to the kernel upgrading to NFTables with IPTables compatibility, and that probably has a different module name), and also a PR for the systemd unit dependencies you suggested.

Will do.
